### PR TITLE
🐛 Delete bastion if basion.enabled=false

### LINKED ISSUE #647: Bastion remains once created
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

- Deployed a cluster with bastion by the manifest `bastion.enabled: true`
- Applied the manifest  `bastion.enabled: false`
- Bastion remains

**What did you expect to happen:**

I expected bastion is deleted by  `bastion.enabled: false`

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Cluster API Provider OpenStack version (Or `git rev-parse HEAD` if manually built): 89b3519fbe56128d0f056323d3027ec52f53e48d
- Cluster-API version: 0.3.9
- Minikube/KIND version: 0.7.0
- Kubernetes version (use `kubectl version`): 1.18.9
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by hidekazuna**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #647 

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**TODOs**:

- [ ] squashed commits
- if necessary:
  - [ ] includes documentation
  - [ ] adds unit tests

/hold

### CODE REVIEW

File: pkg/cloud/services/compute/bastion.go [hunk 1]
```diff
@@ -23,15 +23,20 @@ import (
 	"sigs.k8s.io/cluster-api-provider-openstack/pkg/record"
 )
 
-func (s *Service) DeleteBastion(serverID string) error {
+func (s *Service) DeleteBastion(serverID string, openStackCluster *infrav1.OpenStackCluster) error {
```
Comment (sbueringer) [inline pkg/cloud/services/compute/bastion.go:26]: I would hand over the cluster as first parameter for consistency

File: pkg/cloud/services/networking/securitygroups.go [hunk 2]
```diff
@@ -375,14 +375,18 @@ func (s *Service) generateDesiredSecGroups(secGroupNames map[string]string, open
 	return desiredSecGroups, nil
 }
 
-func (s *Service) DeleteSecurityGroups(group *infrav1.SecurityGroup) error {
+func (s *Service) DeleteSecurityGroups(group *infrav1.SecurityGroup, openStackCluster *infrav1.OpenStackCluster) error {
```
Comment (sbueringer) [inline pkg/cloud/services/networking/securitygroups.go:378]: I would hand over the cluster as first parameter

File: controllers/openstackcluster_controller.go [hunk 3]
```diff
@@ -236,6 +209,36 @@ func contains(arr []string, target string) bool {
...
+	computeService, err := compute.NewService(osProviderClient, clientOpts, log)
+	if err != nil {
+		return err
+	}
+	networkingService, err := networking.NewService(osProviderClient, clientOpts, log)
+	if err != nil {
+		return err
+	}
+
+	if bastion := openStackCluster.Status.Bastion; bastion != nil {
```
Comment (sbueringer) [inline controllers/openstackcluster_controller.go:222]: I wouldn't write this in an additional var. That's usually only done for errors or stuff like map lookups in Golang. So please just use `openStackCluster.Status.Bastion`

**Review by sbueringer — CHANGES_REQUESTED** [review]
overall lgtm, just a few nits

**Comment by hidekazuna** [comment]
@sbueringer fixed.

I will propose another PR to update parameter order for other functions.

**Comment by jichenjc** [comment]
/lgtm

the param reorder in next PR works for me

**Comment by sbueringer** [comment]
> @sbueringer fixed.
> 
> I will propose another PR to update parameter order for other functions.

Thx, sounds good to me. I think the following is a good ordering (just an example for when we need *all* of the commonly used parameters):
```go
(ctx context.Context, log logr.Logger, ctrlClient, workloadClient client.Client, kubeClient, workloadKubeClient kubernetes.Interface, patchHelper *patch.Helper, cluster *clusterv1.Cluster, openStackCluster *infrav1.OpenStackCluster, machine *clusterv1.Machine, openstackMachine *infrav1.OpenStackMachine, .... (everything else)) (ctrl.Result, error) 
```
Some background why:
* context first
* then logr.Logger
* first clients for the management cluster, then the ones for the workload clusters
  * sorted per type: we need the controller runtime client more then the regular one (not sure if we need the regular one at all right now, in this repo)
* patchHelper next (also some sort of "client", but not needed that much, and it cannot do that much, so in some way a "smaller" client)
* after that it's just from "bigger" / "more important" resources to smaller ones, and first from the cluster-api repo and then ours
* I would continue this "top down" approach for other parameters (although I don't necessarily would enforce a hard ordering for "everything else")

**Review by sbueringer — APPROVED** [review]

**Comment by hidekazuna** [comment]
Let's merge this.

/hold cancel

**Pull request merged**